java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java

```diff
@@ -0,0 +1,297 @@
+  // type.
+  private static final String UNPARSED_FIELD_VALUE =
+      "$$$UNPARSED$$$9A1B0050-1BBD-420F-B6B3-AA0DB7D0AAE8";
+
+  private static final String TRUE_STR = "true";
+  private static final String FALSE_STR = "false";
+
+  /**
+   * testAnnotationSpecs is a map where key is the annotation name and the value is an object
+   * encoding the specification of a testing annotations understood by Piranha, including how to
+   * parse: a) the flag being tested, b) whether the test is requesting the flag in treatment or
+   * control mode, c) (optionally) whether a particular treatment group is being specified
+   */
+  private final ImmutableMultimap<String, TestAnnotationSpecRecord> testAnnotationSpecs;
+
+  // Instantiate using Builder.
+  private TestAnnotationResolver(
+      ImmutableMultimap<String, TestAnnotationSpecRecord> testAnnotationSpecs) {
+    this.testAnnotationSpecs = testAnnotationSpecs;
+  }
+
+  // Pre-parse the actual arguments of an annotation instance into a multi-map of
+  // "$fieldName" -> {val1, val2, ...}, before trying to match to an spec.
+  private static ImmutableMultimap<String, AnnotationArgument> parseAnnotationArguments(
+      AnnotationTree at) {
+    ImmutableMultimap.Builder<String, AnnotationArgument> builder = ImmutableMultimap.builder();
+    for (ExpressionTree et : at.getArguments()) {
+      if (et.getKind() == Tree.Kind.ASSIGNMENT) {
+        AssignmentTree assn = (AssignmentTree) et;
+        String key = "$" + ASTHelpers.getSymbol(assn.getVariable()).getSimpleName().toString();
+        ExpressionTree assnExpression = assn.getExpression();
+        Tree.Kind assnExprKind = assnExpression.getKind();
+        switch (assnExprKind) {
+          case IDENTIFIER: // Fallthrough
+          case MEMBER_SELECT:
+            builder.put(
+                key,
+                new AnnotationArgument(
+                    ASTHelpers.getSymbol(assnExpression).getSimpleName().toString(),
+                    assnExpression));
+            break;
+          case STRING_LITERAL: // Fallthrough
+          case BOOLEAN_LITERAL:
+            builder.put(
+                key,
+                new AnnotationArgument(
+                    ((LiteralTree) assnExpression).getValue().toString(), assnExpression));
+            break;
+          case NEW_ARRAY:
+            // For each in the array
+            NewArrayTree arrayExpression = (NewArrayTree) assnExpression;
+            for (ExpressionTree expr : arrayExpression.getInitializers()) {
+              if (expr instanceof LiteralTree) {
+                builder.put(
+                    key, new AnnotationArgument(((LiteralTree) expr).getValue().toString(), expr));
+              } else {
+                builder.put(
+                    key,
+                    new AnnotationArgument(
+                        ASTHelpers.getSymbol(expr).getSimpleName().toString(), expr));
```

Comment:
Can we have a helper for ASTHelpers.getSymbol(expr).getSimpleName().toString()?

Reply:
Sure. This is fairly idiomatic EP, but I also don't see the harm in it :)